src/main/java/engineering/swat/watch/impl/BundledSubscription.java

```diff
@@ -52,51 +54,58 @@ public void accept(R t) {
-            // we just added a new one
-            // so lets take a lock on it, and try to be the one that gets to initialize it
+        while (true) {
+            Subscription<Event> active = this.subscriptions.computeIfAbsent(target, t -> new Subscription<>());
+            // after this, there will only be 1 instance of active subscription in the map.
+            // but we might have a race with remove, which can close the subscript between our get and our addition
+            // since this code is very hard to get right without locks, and shouldn't be run too often
+            // we take a big lock around the subscription management
             synchronized(active) {
-                // now lock on it to make sure nobo
-                if (active.toBeClosed == null) {
-                    first = true;
-                    active.add(eventListener); // we know we already have the lock, and we need to do this before we register the watch
-                    var newSubscriptions = wrapped.subscribe(target, active);
-                    active.setToBeClosed(newSubscriptions);
+                if (active.closed) {
+                    // we lost the race with closing the subscription, so we retry
+                    continue;
                 }
-                else {
+                active.add(eventListener);
+                if (active.toBeClosed == null) {
+                    // the watch is not active yet, and we were the first to get the lock
+                    active.toBeClosed = wrapped.subscribe(target, active);
                 }
             }
-        }
-        // at this point we have to be sure that we're not the first to in the list
-        // since we might have won the race on the compute, but lost the race
-        if (!first) {
-            active.add(eventListener);
-        }
-        return () -> {
-            active.remove(eventListener);
-            if (!active.hasActiveConsumers()) {
-                subscriptions.remove(target);
-                if (active.hasActiveConsumers()) {
-                    // we lost the race, someone else added something again
-                    // so we put it back in the list
-                    subscriptions.put(target, active);
+            return () -> {
+                boolean scheduleClose = false;
+                synchronized(active) {
+                    active.remove(eventListener);
+                    scheduleClose = !active.hasActiveConsumers() && !active.closed;
                 }
-                else {
-                    if (active.toBeClosed != null) {
-                        active.toBeClosed.close();
-                    }
+                if (scheduleClose) {
+                    // to avoid hammering the system with closes & registers in a short periode
+                    // we schedule the cleanup of watches in the background, when even after a small delay
+                    // nobody is interested in a certain file anymore
```

Comment:
Oh... So that's what you've been up to yesterday afternoon 😉

Reply:
yes, but this actually solved some very hard to reproduce races in the JDK library itself. 

and it also hammers the jdk & kernel a bit less, which is good.